include/kodlab_mjbots_sdk/pid_controller.h

```diff
@@ -0,0 +1,648 @@
+
+  /**
+   * @brief Function to get Kd gain 
+   */
+  VectorNS get_kd_gain(){return kd_;}
+  
+  /**
+   * @brief Function to set the scalar deadband
+   * @param deadband Deadband that needs to be set
+   */
+  void set_deadband(Scalar deadband) { deadband_ = deadband*VectorNS::Ones(); }
+
+  /**
+   * @brief Function to set the vector deadband
+   * @param deadband Deadband that needs to be set
+   */
+  void set_deadband(const VectorNS & deadband) { deadband_ = deadband; }
+
+  /**
+   * @brief Function to get deadband 
+   */
+  VectorNS get_deadband(){return deadband_;}
+
+
+  /**
+   * @brief Function to set the scalar accumulator limit
+   * @param limit accumulator limit
+   */
+  void set_accumulator(Scalar limit){ accumulator_limit_ = limit*VectorNS::Ones();}
+
+  /**
+   * @brief Function to set the vector accumulator limit
+   * @param limit accumulator limit
+   */
+  void set_accumulator(const VectorNS & limit){ accumulator_limit_ = limit;}
+
+
+  /**
+   * @brief Function to set the Saturation
+   * @param saturation Saturation
+   */
+  void set_saturation_limits(std::array<VectorNS, 2> saturation) {
+    saturation_ = saturation;
+  }
+
+  
+  void set_saturation_limits(VectorNS saturation) {
+    std::array<VectorNS, 2> out = {-saturation, saturation};
+    set_saturation_limits(out);
+  }
+
+  void set_saturation_limits(std::array<Scalar,2> saturation) {
+    set_saturation_limits({saturation[0]*VectorNS::Ones(), 
+                           saturation[1]*VectorNS::Ones()});
+  }
+
+  void set_saturation_limits(Scalar saturation) {
+    
+    set_saturation_limits({-saturation*VectorNS::Ones(), 
+                            saturation*VectorNS::Ones()});
+  }
+  /**
+   * @brief Function to get the dimensions
+   */
+  int get_N(){ return N;}
+
+  /**
+   * @brief Function to reset the integral error to zero
+   */
+  void reset_i_error() { i_error_ = VectorNS::Zero();}
+
+  /**
+   * @brief Function to update the state setpoints and reset integral error
+   * @param setpoint State Target of the controller.
+   * @param derivative_setpoint Derivative State Target of the controller.
+   */
+  void set_setpoints(const VectorNS & setpoint, 
+                     const VectorNS & derivative_setpoint = VectorNS::Zero()){
+    setpoint_ = setpoint; 
+    derivative_setpoint_ = derivative_setpoint;  
+  }
+
+  /**
+   * @brief Function to update the state setpoints and reset integral error
+   * @param setpoint State Target of the controller.
+   * @param derivative_setpoint Derivative State Target of the controller.
+   */
+  void set_setpoints(Scalar setpoint, Scalar derivative_setpoint = 0){
+    //checking if using for scalar PID
+    static_assert(N==1, "Using Scalar function in vector PID");
+    set_setpoints(setpoint*VectorNS::Ones(), 
+                  derivative_setpoint*VectorNS::Ones());
+  }
+  
+  /**
+   * @brief Function to get State setpoint
+   */
+  VectorNS get_state_setpoint(){return setpoint_;}
```

Comment:
Thoughts on using copy to return a potentially large vector? @jdcaporale This is used in a number of getters.

Reply:
Agreed. Not trying to give access though, so make the output a  const reference.